application/apps/indexer/session/src/state/session_file.rs

```diff
@@ -0,0 +1,279 @@
+                        e
+                    )),
+                })?;
+                filename
+            };
+            debug!("Session file setup: {}", filename.to_string_lossy());
+            self.filename = Some(filename.clone());
+            Ok(Grabber::lazy(TextFileSource::new(&filename))
+                .map(|g| self.grabber = Some(Box::new(g)))?)
+        } else {
+            Ok(())
+        }
+    }
+
+    #[allow(clippy::len_without_is_empty)]
+    pub fn len(&mut self) -> Result<usize, NativeError> {
+        if let Some(ref grabber) = self.grabber {
+            if let Some(md) = grabber.get_metadata() {
+                Ok(md.line_count)
+            } else {
+                Ok(0)
+            }
+        } else {
+            Ok(0)
+        }
+    }
+
+    pub async fn write(
+        &mut self,
+        source_id: u8,
+        state_cancellation_token: CancellationToken,
+        msg: String,
+    ) -> Result<SessionFileState, NativeError> {
+        if self.writer.is_none() && self.filename.is_none() {
+            return Err(NativeError {
+                severity: Severity::ERROR,
+                kind: NativeErrorKind::Grabber,
+                message: Some(String::from("Session file isn't assigned yet")),
+            });
+        }
+        if self.writer.is_none() {
+            if let Some(ref filename) = self.filename {
+                self.writer = Some(BufWriter::new(File::create(filename).map_err(|e| {
+                    NativeError {
+                        severity: Severity::ERROR,
+                        kind: NativeErrorKind::Io,
+                        message: Some(format!(
+                            "Fail to create session writer for {}: {}",
+                            filename.to_string_lossy(),
+                            e
+                        )),
+                    }
+                })?));
+            }
+        }
+        let writer = &mut (self.writer.as_mut().ok_or(NativeError {
+            severity: Severity::ERROR,
+            kind: NativeErrorKind::Grabber,
+            message: Some(String::from("Session file writer wasn't created")),
+        })?);
+        writer.write(msg.as_bytes()).map_err(|e| NativeError {
+            severity: Severity::ERROR,
+            kind: NativeErrorKind::Io,
+            message: Some(e.to_string()),
+        })?;
+        if self.observing.is_source_changed(source_id)
+            || self.last_message_timestamp.elapsed().as_millis() > NOTIFY_IN_MS
+        {
+            writer.flush().map_err(|e| NativeError {
+                severity: Severity::ERROR,
+                kind: NativeErrorKind::Io,
+                message: Some(e.to_string()),
+            })?;
+            self.last_message_timestamp = Instant::now();
+            self.update(source_id, state_cancellation_token).await
+        } else {
+            self.has_updates = true;
+            Ok(SessionFileState::MaybeChanged)
+        }
```

Comment:
there is a way to write that without having to re-check if the writer really exists. 
``` rust
        let writer = match (&mut self.writer, &self.filename) {
            (None, None) => {
                return Err(NativeError {
                    severity: Severity::ERROR,
                    kind: NativeErrorKind::Grabber,
                    message: Some(String::from("Session file isn't assigned yet")),
                });
            }
            (None, Some(ref filename)) => {
                let writer = BufWriter::new(File::create(filename).map_err(|e| NativeError {
                    severity: Severity::ERROR,
                    kind: NativeErrorKind::Io,
                    message: Some(format!(
                        "Fail to create session writer for {}: {}",
                        filename.to_string_lossy(),
                        e
                    )),
                })?);
                self.writer.insert(writer)
            }
            (Some(writer), None) => writer,
            (Some(writer), Some(_)) => writer,
        };
        writer.write_all(msg.as_bytes())?;
        if self.observing.is_source_changed(source_id)
            || self.last_message_timestamp.elapsed().as_millis() > NOTIFY_IN_MS
        {
            writer.flush()?;
            self.last_message_timestamp = Instant::now();
            self.update(source_id, state_cancellation_token).await
        } else {
            self.has_updates = true;
            Ok(SessionFileState::MaybeChanged)
        }
```